### DISCUSSION TITLE
Problem with calculation result

### DISCUSSION THREAD

**Discussion by Chaoyue-Liu** [idaholab/moose#23258] [Q&A General] [answered]
Dear Moose Team,
I built a model for lithium ion battery electrochemistry, and I got this wierd problem. You can find the header files, source files, and input files in here [https://github.com/Chaoyue-Liu/model](https://github.com/Chaoyue-Liu/model).
If you 3D_full_model.i, even though it converge really quick, the result doesn't make sense. Especially if you change the phis_neumann boundary condition value from negative to positive, the result doesn't change at all. But if I delete the Volum3_TET4 domain at the bottom, which is a current conductor with equation of ohm's law, the result is right. I wonder why this is happening, could you check it? Many thanks!

**Comment by GiudGiud** [comment]
Hello

I looked at your mesh and there's very high ratios between the element volumes. That could have been an issue. 
Can you attach the convergence log please?

Guillaume

**Reply by Chaoyue-Liu** [reply]
Hello,
I regenerated the meshes with a lower element volume ratio and run the simulations. The screenshots of results and convergence log are attached in the problem folder [https://github.com/Chaoyue-Liu/model](https://github.com/Chaoyue-Liu/model). The issue is still the same. 
1. For both simulation with or without current conductor, they cannot proceed to the end of calculation. 
2. With the same boundary condition, adding the current conductor will cause the calculation totally inversed. Also changing the sign of the boundary condition doesn't take any effect on the result as you can see in the screenshots.
3. The model no_cc (no current conductor) can calculate the correct initial potential of phis=4.2, but the model with cc cannot, the initial phis=3.9 which is just my input initial value
4. For the model no_cc, the output terminal_potential, which you can find in the log, seems turbulating instead of monotonically decreases as cs_integral increases
Thank you!

**Reply by GiudGiud** [reply]
Hello

Which log in particular should I look at?
I looked at this one:
[log](https://github.com/Chaoyue-Liu/model/blob/main/problem/log_full_model%20phis_neumann%3D-10.txt)

It seems the linear solve does not converge. You can see it stabilize at a given residual and stop improving. 
Please have a look at using a different preconditioner
https://mooseframework.inl.gov/syntax/Preconditioning/
This has more on preconditionig (this is an executioner though, but these info on preconditioning are valid outside of it)
https://mooseframework.inl.gov/source/executioners/Steady.html

Guillaume

**Reply by Chaoyue-Liu** [reply]
Hello
I'll look into the preconditioner, thank you. I hope that you could take a look at both two models so I attached the two logs. The biggest problem of the model is not just convergence, but also the problem that adding the current conductor domain(which only applies a laplacian equation for static potential calculation) will totally change the calculation result, and the sign of neumann boundary condition doesn't take effect. Could you check what is causing this?

**Reply by GiudGiud** [reply]
if the solution is not converged, the results can be whatever. There is no guarantee. I think we should focus on getting a clean linear then non linear convergence.

**Reply by Chaoyue-Liu** [reply]
I see. I will try the different preconditioners to see if I can get a better linear convergence. Thanks!

**Reply by Chaoyue-Liu** [reply]
Hello Guillaume, I've got it run properly by changing the preconditioner. Many thanks! There is still one problem.
![image](https://user-images.githubusercontent.com/95432912/216049589-cb45d267-697d-4581-bdbf-c0bedbb75d53.png)
Here are the domains that calculate the variable phis by ADMatDiffusion. The diffusivity of the bottom blue cube should 3.5e7, which is much higher than the rest, which is of the magnitude of 0.1 or 1. This will cause a calculation fail to converge. But if I change 3.5e7 to a value of much smaller magnitude, the model is able to run. Is there any way to make it possible to run with value of 3.5e7?
The following is the error message:
```
Framework Information:
MOOSE Version:           git commit 58950ae311 on 2022-09-22
LibMesh Version:
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Wed Feb  1 14:05:00 2023
Executable Timestamp:    Wed Feb  1 11:06:28 2023

Parallelism:
  Num Processors:          4
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:
    Total:                 30399
    Local:                 7790
    Min/Max/Avg:           6820/8021/7599
  Elems:
    Total:                 163571
    Local:                 40893
    Min/Max/Avg:           40892/40893/40892
  Num Subdomains:          5
  Num Partitions:          4
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                76287
  Num Local DOFs:          19153
  Variables:               "phis" { "phil" "cl" } "cs"
  Finite Element Types:    "LAGRANGE" "LAGRANGE" "LAGRANGE"
  Approximation Orders:    "FIRST" "FIRST" "FIRST"

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP (auto)

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.


    Building SemiLocalElemMap                                                            [  5.25 s] [    0 MB]
  Finished Performing Initial Setup                                                      [  9.48 s] [    1 MB]
Still Executing
Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+--------------------+
| time           | cs_integral    | terminal_potential |
+----------------+----------------+--------------------+
|   0.000000e+00 |   0.000000e+00 |       0.000000e+00 |
+----------------+----------------+--------------------+


Time Step 1, time = 10, dt = 10

Performing automatic scaling calculation

 0 Nonlinear |R| = 2.549927e+02
      0 Linear |R| = 2.549927e+02
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_OUTMEMORY
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
  Finished Solving                                                                       [ 20.82 s] [   86 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.

Time Step 1, time = 5, dt = 5
 0 Nonlinear |R| = 2.549927e+02
      0 Linear |R| = 2.549927e+02
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_OUTMEMORY
```

**Reply by GiudGiud** [reply, marked as answer]
You ll want to look at preconditioning again
What s happening is that the conditioning of your matrix is likely becoming poor as the coeficients in part of the system are becoming much larger than in others.

You could try to separate the variables there, using interface kernels to make the connection between the sets of variables, and using automatic scaling to scale the equations. 

I m afraid we dont have spatially dependent preconditioning
